File: 10/partone.js
const path = require('path');
const filePath = path.join(__dirname, 'input.txt');
const readFile = require('../utils').readFile;

// Processing the input into a suitable data structure
function processInput(input) {
  return input.split('\n').map((number) => parseInt(number, 10));
}

async function main() {
  const input = await readFile(filePath);
  const adapters = processInput(input).sort((a, b) => (a > b ? 1 : -1));

  let distribution = { 1: 0, 2: 0, 3: 1 };
  let prevJoltage = 0;

  for (adapter in adapters) {
    const joltage = adapters[adapter];
    const diff = joltage - prevJoltage;

    distribution[diff] += 1;
    prevJoltage = joltage;
  }
  console.log(distribution[1] * distribution[3]);
}

main();

File: 2021/07/part-one.rb
def read_file
  file = File.open('input.txt')
  file.readline.split(",").map(&:to_i)
end

def median(array)
  return nil if array.empty?
  sorted = array.sort
  len = sorted.length
  (sorted[(len - 1) / 2] + sorted[len / 2]) / 2.0
end

def main
  data = read_file
  median = median(data)
  total = 0

  for i in data do
    total += (median - i).abs
  end

  print total
end

puts main
